### PR TITLE
🎉 New Source: UK Police

### PR DESCRIPTION
**Pull request by isaacharrisholt**
## What

Fraud detection is becoming a massive part of many industries, especially FinTech. The UK has a free, open police database with a JSON API. It's handy to be able to use this data frequently.

## How

This PR adds a UK Police source with streams regarding UK police forces, crimes and stop and search operations.

## Recommended reading order

1. `uk_police.yaml`
2. `spec.yaml`

## 🚨 User Impact 🚨

No changes to existing code.

## Pre-merge Checklist

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

None.

</details>

<details><summary><strong>Integration</strong></summary>

image

</details>

<details><summary><strong>Acceptance</strong></summary>

image

</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
Thanks for the contribution! Someone will review it during this week. If you have any questions please create comments so the reviewer can focus on those points.

File: airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml:25]: Rename this parameter, it is not a month. start_year_month

File: airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml:55]: Why you don't make those streams substream of force one and retrieve all data to all forces?

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml:55]: Similar to the MonthStreamSlicer you can create one for Forces.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Comment by marcosmarxm** [comment]
Some comments, please add the documentation page too.

File: airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml:25]: Done

File: airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml:55]: I could, but would that not mean I wouldn't be able to use the `DatetimeStreamSlicer` for those streams? I don't believe I can use both

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml:55]: @girarda is it possible to interpolate two StreamSlicer?

**Comment by marcosmarxm** [comment]
Thanks @isaacharrisholt I asked to Alex some help about the stream slicer.

File: airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml
Comment (girarda) [reply airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml:55]: @isaacharrisholt you can wrap in a [cartesian product stream slicers](https://docs.airbyte.com/connector-development/config-based/understanding-the-yaml-file/stream-slicers#cartesian-product-stream-slicer) to iterate over all the pairs (force, datetime)

Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml]: @girarda I've made that change, but I'm getting the following when I run the check operation:

```
{"type": "LOG", "log": {"level": "ERROR", "message": "Check failed"}}
{"type": "CONNECTION_STATUS", "connectionStatus": {"status": "FAILED", "message": "\"Unable to connect to stream crimes_with_no_location - 'NoneType' object has no attribute 'get'\""}}
```

I've pushed the changes to the repo in commit [2664eb8](https://github.com/airbytehq/airbyte/pull/18344/commits/2664eb8903b139e109db2efd4e3852e2bbe0401b)

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml]: @girarda can you check this change?

**Comment by marcosmarxm** [comment]
@isaacharrisholt let's wait for Alex comments about the StreamSlicer, tomorrow I'll check again.

File: airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml
Comment (girarda) [reply airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml]: found the [bug here](https://github.com/airbytehq/airbyte/pull/18529/files#diff-9328dc3f2f6ae4ce085ddc1e7da60a9c02002e64d9311ca907f07abf411409d4R89)

@isaacharrisholt to unblock yourself, you can add do the same None check and use your local version of the CDK.

you can do this by running the following commands from the connector's directory:
```
pip uninstall airbyte_cdk
pip install -e ../../../airbyte-cdk/python/
```

I tried reading from the stream, but I'm getting a 404, but I also get reach the sample request from their doc (https://data.police.uk/api/crimes-no-location?category=all-crime&force=leicestershire&date=2017-03)

Let me know if you have more questions!

Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-uk-police/source_uk_police/spec.yaml]: @girarda I'm also getting a 404, but it seems like the `force` request parameter isn't being properly injected.

```
{"type": "DEBUG", "message": "Making outbound API request", "data": {"request_body": "None", "headers": "{'User-Agent': 'python-requests/2.28.1', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive'}", "url": "https://data.police.uk/api/crimes-no-location?lat=51.5074&lng=-0.1278&category=burglary&date=2022-08"}}
```

This could be an issue with my configuration, but I don't believe so?

**Comment by marcosmarxm** [comment]
@isaacharrisholt did you try with latest version of the low-code CDK? https://github.com/airbytehq/airbyte/pull/18529

**Comment by isaacharrisholt** [comment]
> @isaacharrisholt did you try with latest version of the low-code CDK? #18529

@marcosmarxm yep, still getting the 404 due to the param not being injected.

**Comment by marcosmarxm** [comment]
Hello! I'm going to be out of the office this Friday and won't be able to review your contribution again today, I return next Monday. So far, most contributions look solid and are almost done to be approved. As said in [Chris' comment](https://airbytehq.slack.com/archives/C0450PCSDB2/p1667430678866629) all contributions made before 2-November are eligible to receive the prize and have 2 weeks to merge the contributions. But I ensure next week we're going to have your contribution merged. If you have questions about the implementation you can send them in `#hacktoberfest-2022` in Airbyte's Slack.

Sorry the inconvenience and see you again next week, thank you so much for your contribution!

**Comment by isaacharrisholt** [comment]
Hey, is there any update on this? Keen to get it merged before the deadline passes :)

**Comment by marcosmarxm** [comment]
Because of missing features in low-code CDK we're not going to merge this contribution now but we'll accept as a valid contribution to receive the prize. Thank you so much for the contribution and I'll return as soon we have an update about the bug.
Please enter in touch with @RealChrisSean :tada:

**Comment by RealChrisSean** [comment]
This has been added to the list

**Comment by marcosmarxm** [comment]
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

**Pull request closed without merging**
